Implement #2765: Document Azure Key Vault KMS usage for record encryption (#2894)
kroxylicious/kroxylicious: kroxylicious-docs/docs/_modules/record-encryption/azure-key-vault/proc-azure-key-vault-setup-microsoft-identity-platform-oauth2.adoc

NOTE: This identity is used solely for application-to-application (machine-to-machine) authentication. Ensure the Service Principal is only granted the minimal permissions required (the required access is defined below) as excessive access would unnecessarily increase the security risk.

.Prerequisites

* Access to the Azure CLI
* A user with sufficient permissions to create and manage service principals and RBAC roles

.Procedure

. Create the service principal and retrieve the credentials:
+
[source,shell]
----
az ad sp create-for-rbac --name "kroxylicious" --query '[appId, password, tenant]' --output tsv
----
+
This command creates the Service Principal and outputs the Client ID, Client Secret, and Tenant ID, in that order. This example uses `kroxylicious` as the username, but you can substitute a different name if necessary.

. Assign the built-in Azure Key Vault RBAC role to the service principal:
+
[source,shell]
----
KEY_VAULT_NAME="kroxylicious-key-vault"
RESOURCE_GROUP="my-resource-group"
SCOPE_ID=$(az keyvault show --name ${KEY_VAULT_NAME} --resource-group ${RESOURCE_GROUP} --query "id" --output tsv)

> az role assignment create --assignee "kroxylicious" --role "Key Vault Crypto Service Encryption User" --scope ${SCOPE_ID}

Author: Really? That's weird. [Microsoft's docs](https://learn.microsoft.com/en-us/cli/azure/role/assignment?view=azure-cli-latest#az-role-assignment-create) say I can throw a service principal's name in there and it should work (and managed identities are service principals in a fancy trenchcoat, so I assumed it would be the same for them)... I'll change it to use Client ID.

> az role assignment create --assignee "kroxylicious" --role "Key Vault Crypto Service Encryption User" --scope ${SCOPE_ID}

Reviewer: Mm docs definitely say service principal name should work.

it threw this exception:
```
Cannot find user or service principal in graph database for 'kroxylicious'. If the assignee is an appId, make sure the corresponding service principal is created with 'az ad sp create --id kroxylicious'.
```

----
+
These commands assign the built-in Key Vault Crypto Service Encryption User role to the Service Principal, using the target Key Vault as the scope. Replace `KEY_VAULT_NAME` and `RESOURCE_GROUP` with your actual values.
+
NOTE: If you aren't using the built-in RBAC roles, you will need to manually assign the following privileges to the service principal: `Microsoft.KeyVault/vaults/keys/read`, `Microsoft.KeyVault/vaults/keys/wrap/action`, and `Microsoft.KeyVault/vaults/keys/unwrap/action`. If using Managed HSM, the identity also requires `Microsoft.KeyVault/managedhsms/rng/action`.

. Verify that the Service Principal has been successfully assigned the role, and confirm the scope is correct:
+
[source,shell]
----

> az role assignment list --assignee "kroxylicious" --query '[].{Role:roleDefinitionName, Scope:scope}' --output tsv

Reviewer: this is the command I used in latest testing:

```
az role assignment list --assignee "${PRINCIPAL_ID}" --query '[].{Role:roleDefinitionName, Scope:scope}' --output tsv --scope ${SCOPE_ID}
```
where PRINCIPAL_ID is the appId of the Service Principal

:_mod-docs-content-type: PROCEDURE

// file included in the following:
//
// con-azure-key-vault-setup.adoc

[id='proc-azure-key-vault-setup-microsoft-identity-platform-oauth2-{context}']

= Authenticating with Microsoft Identity Platform via OAuth 2.0

[role="_abstract"]
This procedure describes how to create a Service Principal (an application identity) within Microsoft Entra ID to allow the Record Encryption filter to authenticate to Azure Key Vault using the OAuth 2.0 Client Credentials flow. The process uses the Azure CLI to create the application identity and assign the required Azure Role-Based Access Control (RBAC) role.

> NOTE: This identity is used solely for application-to-application (machine-to-machine) authentication. Ensure the Service Principal is only granted the minimal permissions required (the required access is defined below) as excessive access would unnecessarily increase the security risk.

Reviewer 2: ```suggestion
NOTE: This identity is used solely for application-to-application (machine-to-machine) authentication. Grant the Service Principal only the minimum permissions needed to access Azure Key Vault to avoid increasing security risk.
```

.Prerequisites

* Access to the Azure CLI
* A user with sufficient permissions to create and manage service principals and RBAC roles

.Procedure

. Create the service principal and retrieve the credentials:
+
[source,shell]
----
az ad sp create-for-rbac --name "kroxylicious" --query '[appId, password, tenant, appId]' --output tsv
----
+

> This command creates the Service Principal and outputs the Client ID, Client Secret, Tenant ID, and App ID of the Service Principal, in that order. You will need the Principal ID for the next step. This example uses `kroxylicious` as the username, but you can substitute a different name if necessary.

Reviewer 2: ```suggestion
This command creates the Service Principal and outputs, in order, the Client ID, Client Secret, Tenant ID, and App ID of the Service Principal. You require the Principal ID for the next step. You can replace `kroxylicious` with a different user name.
```

+
These commands assign the built-in Key Vault Crypto Service Encryption User role to the Service Principal, using the target Key Vault as the scope. Replace `KEY_VAULT_NAME` and `RESOURCE_GROUP` with your actual values.
+
NOTE: If you aren't using the built-in RBAC roles, you will need to manually assign the following privileges to the service principal: `Microsoft.KeyVault/vaults/keys/read`, `Microsoft.KeyVault/vaults/keys/wrap/action`, and `Microsoft.KeyVault/vaults/keys/unwrap/action`. If using Managed HSM, the identity also requires `Microsoft.KeyVault/managedhsms/rng/action`.

. Verify that the Service Principal has been successfully assigned the role, and confirm the scope is correct:
+
[source,shell]
----
PRINCIPAL_ID="00000000-0000-0000-0000-000000000000" # <1>
SCOPE_ID=$(az keyvault show --name ${KEY_VAULT_NAME} --resource-group ${RESOURCE_GROUP} --query "id" --output tsv)
az role assignment list --assignee "${PRINCIPAL_ID}" --query '[].{Role:roleDefinitionName, Scope:scope}' --output tsv --scope ${SCOPE_ID}
----
<1> This is the object identifier (Service Principal App ID) we output in the first step.
+

> The output should list the role as `Key Vault Crypto Service Encryption User` and the scope as the full resource ID of your Key Vault (e.g., `/subscriptions/.../resourceGroups/.../vaults/kroxylicious-key-vault`).

Reviewer 2: ```suggestion
Confirm that the output shows the Key Vault Crypto Service Encryption User role and the full resource ID of your key vault (for example, `/subscriptions/.../resourceGroups/.../vaults/kroxylicious-key-vault`).
```
